modules/MvwmIconMan/MvwmIconMan.1.in
.\" t
.\" @(#)@PACKAGE@-@VERSION@ @RELDATELONG@
.de EX          \"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH MvwmIconMan 1 "@RELDATELONG@ (@VERSION@)" Mvwm "Mvwm Modules"
.UC
.SH NAME
MvwmIconMan \- an mvwm icon manager
.SH SYNOPSIS
MvwmIconMan is spawned by mvwm, so no command line invocation will work.

.SH DESCRIPTION
MvwmIconMan is an icon manager modeled after the TWM icon manager.
The user may have multiple icon managers, each of which armed with
a list of window types which it manages. For example, the user may
have one manager which lists only emacs windows, and another which
lists everything else. You may also specify what resolution each
icon manager uses, for example, one icon manager may manage
windows on all desks, and another may manage only those on the
current desk, page or screen. MvwmIconMan can display the
miniature icons provided by mvwm for its managed windows. The
managers may have a maximum number of columns (and so grows
vertically), a maximum number of rows (and then grows
horizontally), or stay at a fixed size, and adjust the size of the
window buttons to fit (think win95's Taskbar). And when support is
compiled in for the X Shape extension, then the manager windows
may be shaped.

You can specify actions to be run when mouse, or key events are received. For
example, you could bind the first mouse button to iconify the selected window,
and make bindings for the arrow keys to navigate the manager window without
the mouse.

MvwmIconMan can be set to display which window currently has the keyboard
focus, and by binding the select event (see below) to the mvwm Focus function,
you can emulate the TWM icon manager's behavior.

.SH INITIALIZATION
During initialization, MvwmIconMan searches though the mvwm configuration file
for the options which are described below. It is highly recommended that you
make MvwmIconMan be a sticky window. And if you want to make use of the
followfocus option, and/or binding an action to Focus, then you should make
MvwmIconMan clicktofocus. Also, when using the Shape option, it's recommended
that the MvwmIconMan window not be decorated at all by mvwm.

.SH INVOCATION
MvwmIconMan can be invoked by inserting the line 'Module MvwmIconMan' in
the .mvwm2rc file.  If MvwmIconMan is to be spawned during mvwm's
initialization, then this line should be placed in the StartFunction
declarations, or it can be bound to a menu, mouse button, or keystroke to
invoke it later.

If you wish to run MvwmIconMan in a transient mode, such as with the built in
window list, then pass "-Transient" as an argument. The invocation "Module
MvwmIconMan -Transient" will do nicely. In this mode, MvwmIconMan will pop up
one manager window directly under the cursor. When the mouse button is
released, it will execute the appropriate action, and then exit.  Things are
somewhat complicated by the fact that you can specify that MvwmIconMan creates
multiple manager windows, behavior which is unsuitable when running
transiently. So, when running transiently, MvwmIconMan will only create one
manager window. Use the manager id 'transient' to specify options for this
manager window.

MvwmIconMan may accept an alias name as an argument.
For example, "Module MvwmIconMan MvwmIconMan-Variant2".

.SH CONFIGURATION OPTIONS REFERENCE CHART
MvwmIconMan has acquired quite a few options. I assume others
share my dislike of paging though a long man page, so here is a
terse reference chart describing the available options. They are
described in more detail in the next section.

.ft C                   \" Courier
.nf
Name            Description                Default
.ft P

NumManagers     number of managers         1
Action          binds command to event     Mouse 0 N sendcommand Iconify
Background      default background         gray
ButtonGeometry  size of button in pixels
Colorset        default colorset
DontShow        list of windows to ignore
DrawIcons       use mini icons             false
FocusAndSelectButton                       flat grey black
FocusAndSelectColorset
FocusButton     style for focused buttons  up grey black
FocusColorset
FollowFocus     show which win has focus   false
Font                                       8x13
Foreground      default text color         white
Format          describes button label     "%c: %i"
IconName        manager icon name          MvwmIconMan
IconAndSelectButton                        up black grey
IconAndSelectColorset
IconButton      style for icon buttons     up black grey
IconColorset
ManagerGeometry size of manager in buttons 0x1
MaxButtonWidth  max width of a button
MaxButtonWidthByColumns
NoIconAction    animate iconification      NOP
PlainButton     style for normal buttons   up black grey
PlainColorset
ReliefThickness size of button relief      2
Resolution      global/desk/page/screen    page
Reverse         normal, icon or none	   none
SelectButton    style for selected buttons flat black grey
SelectColorset
Shape           use shape extension        false
Show            list of windows to show
ShowOnlyIcons   only icons visible         false
ShowNoIcons     icons are not displayed    false
ShowTransient   transient windows visible  false
ShowOnlyFocused only focused visible       false
Sort            keep managers sorted       name
SortWeight      weight for sorting
Tips            Tool Tips mode             none
TipsDelays      Tool Tips mapping delays   1000 300
TipsFont        Font for Tool Tips         default mvwm font
TipsColorset    Tool Tips Colorset         0
TipsFormat      describes Tips label       the Format value
TipsBorderWidth Tool Tips border size      1
TipsPlacement   Tips placement vs button   updown
TipsJustification Tips Just vs button      leftup
TipsOffsets     Tips placement Offsets     3 2
Title           manager title              MvwmIconMan
TitleButton     style for title button     raisededge black grey
TitleColorset
UseWinList      honor WinListSkip?         true

.fi

.SH CONFIGURATION OPTIONS
With the exception of the nummanagers option, all of the options may be
defined on a per-manager basis. So, for example, the user may have his emacs
manager with a red foreground, and his xterm manager with a blue one. A
configuration line may therefore have one of two forms:

.IP "*MvwmIconMan: OptionName OptionValue"
To specify that the \fIOptionName\fP takes the value \fIOptionValue\fP
for all managers.
.IP "*MvwmIconMan: ManagerId OptionName OptionValue"
To specify that the option \fIOptionName\fP takes the value \fIOptionValue\fP
for manager \fIManagerId\fP. \fIManagerId\fP may either be a positive integer,
or the string "transient". An integer id refers to managers which MvwmIconMan
creates when running normally, and an id of "transient" refers to the single
manager which MvwmIconMan creates when running transiently.

The old syntax, that uses an asterisk instead of white spaces
before \fIManagerId\fP and \fIOptionName\fP, is supported too,
but it is obsolete now.

.PP
The following options may be specified:

.IP "*MvwmIconMan: NumManagers \fInum\fP"
\fInum\fP is a positive integer specifying the total number of icon managers.
Since MvwmIconMan would like to know how many managers there are before
handling any manager specific options, this should come first. The default
is 1.

.IP "*MvwmIconMan: [id] Action \fItype\fP \fIbinding\fP"
Binds an MvwmIconMan command to an event. \fIType\fP may be one of the values:
Key, Mouse, or Select. Actions are described in the following section ACTIONS.

.IP "*MvwmIconMan: [id] Background \fIbackground\fP"
Specifies the default background color.

.IP "*MvwmIconMan: [id] ButtonGeometry \fIgeometry\fP"
Specifies the initial geometry of an individual button in pixels. If the
specified height is 0, then the button height is determined from the font
size. X and Y coordinates are ignored.

.IP "*MvwmIconMan: [id] Colorset \fIcolorset\fP"
The default colorset used. Overrides background and foreground. See MvwmTheme.

.IP "*MvwmIconMan: [id] DrawIcons \fIvalue\fP"
If your version of mvwm is capable of using mini icons, then this option
determines if MvwmIconMan displays the mini icons. Otherwise, it generates
an error message.  "true" means that mini icons are shown for iconified
windows, "false" that mini icons are never shown, and "always" that mini icons
are shown for all windows.

.IP "*MvwmIconMan: [id] FocusAndSelectButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons which are
both selected, and have the keyboard focus.

.IP "*MvwmIconMan: [id] FocusAndSelectColorset \fIcolorset\fP"
Works like focusandselectbutton but uses colorsets instead.  The style setting can
still only be applied with focusandselectbutton.  See MvwmTheme.

.IP "*MvwmIconMan: [id] FocusButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons whose
windows have the keyboard focus.

.IP "*MvwmIconMan: [id] FocusColorset \fIcolorset\fP"
Works like focusbutton but uses colorsets instead.  The style setting can still
only be applied with focusbutton.  See MvwmTheme.

.IP "*MvwmIconMan: [id] FollowFocus \fIboolean\fP"
If \fItrue\fP, then the button appearance reflects
which window currently has focus.  Default is false.

.IP "*MvwmIconMan: [id] Font \fIfont\fP"
Specifies the font to be used for labeling the buttons. The default is 8x13.

.IP "*MvwmIconMan: [id] Foreground \fIforeground\fP"
Specifies the default foreground color.

.IP "*MvwmIconMan: [id] Format \fIformatstring\fP"
A printf like format string which describes the string to be printed in the
manager window for each managed window. Possible flags are: %t, %i, %c, and
%r for the window's title, icon title, class, or resource name, respectively.
The default is "%c: %i". \fBWarning\fP: m4 reserves the word \fIformat\fP,
so if you use m4, take appropriate action.

.IP "*MvwmIconMan: [id] IconName \fIiconstring\fP"
Specifies the window icon name for that manager window. \fIIconstring\fP
may either be a single word, or a string enclosed in quotes. The default is
"MvwmIconMan".

.IP "*MvwmIconMan: [id] IconButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons whose
windows are iconified.

.IP "*MvwmIconMan: [id] IconColorset \fIcolorset\fP"
Works like iconbutton but uses colorsets instead.  The style setting can still
only be applied with iconbutton.  See MvwmTheme.

.IP "*MvwmIconMan: [id] IconAndSelectButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons whose
windows are iconified and the button is selected.

.IP "*MvwmIconMan: [id] IconButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons whose
windows are iconified.

.IP "*MvwmIconMan: [id] IconAndSelectColorset \fIcolorset\fP"
Works like IconAndSelectButton but uses colorsets instead.  The style setting
can still only be applied with iconbutton.  See FvwmTheme.

.IP "*MvwmIconMan: [id] ManagerGeometry \fIgeometry\fP"
Specifies the initial geometry of the manager, in units of buttons. If
\fIheight\fP is 0, then the manager will use \fIwidth\fP columns, and will
grow vertically once it has more than \fIwidth\fP windows. Likewise, if
\fIwidth\fP is 0, it will use \fIheight\fP rows, and grow horizontally.  If
both are nonzero, then the manager window will be exactly that size, and stay
that way. As columns are created, the buttons will narrow to accommodate.  If
the geometry is specified with a negative y coordinate, then the window
manager will grow upwards. Otherwise, it will grow downwards.

.IP "*MvwmIconMan: [id] MaxButtonWidth \fIwidth\fP"
Defines a maximum for the width of a button (in pixels). By default there
is no maximum. A value of 0 resets the default. The maximum is only used
with a non growing manager (the ManagerGeometry option
specifies non zero width and height).

.IP "*MvwmIconMan: [id] MaxButtonWidthByColumns \fIcol\fP"
This is another way to set the button width.
col is the number of columns of icons.
The button width is determined by dividing
the total width of MvwmIconMan
by the number of columns.
For example if the
width of MvwmIconMan manager is 1024, MaxButtonWidthByColumns is 4
then MaxButtonWidth is 256.
This is useful when you do not
know, at config time, the width of the manager, for example,
for a swallowed MvwmIconMan.

.IP "*MvwmIconMan: [id] NoIconAction  \fIaction\fP"
Tells MvwmIconMan to do \fIaction\fP when a NoIcon style window is
iconified or de-iconified. Relevant coordinates are appended to \fIaction\fP so
that the icon can be traced to an MvwmIconMan button. An example action
is "*FvwwmIconMan: NoIconAction SendToModule MvwmAnimate animate". A blank or
null action turns this feature off.

.IP "*MvwmIconMan: [id] PlainButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Specifies how normal buttons look. \fIstyle\fP may be one of \fIflat\fP,
\fIup\fP, \fIdown\fP, \fIraisededge\fP, or \fIsunkedge\fP, and describes how
the button is drawn. The color options are both optional, and if not set, then
the default colors are used. If on a monochrome screen, then the \fIstyle\fP
option is ignored, but must still be set.

.IP "*MvwmIconMan: [id] PlainColorset \fIcolorset\fP"
Works like plainbutton but uses colorsets instead.  The style setting can
still only be applied with plainbutton.  See MvwmTheme.

.IP "*MvwmIconMan: [id] ReliefThickness \fInum\fP"
\fInum\fP is an integer specifying the number of pixels thick
that the relief at the edge of non-flat buttons should be.  Setting
this to 0 will produce flat buttons, as if the values for
\fIFocusAndSelectButton\fP, \fIFocusButton\fP, \fIIconAndSelectButton\fP,
\fIIconButton\fP, \fIPlainButton\fP, \fISelectButton\fP, and \fITitleButton\fP
were all set to \fIflat\fP.  If \fInum\fP is negative, the button will be
inverted as if you had used \fIReverse\fP for all classes.

.IP "*MvwmIconMan: [id] Resolution \fIresolution\fP"
Specifies when the manager will display an entry for a certain
window. \fIresolution\fP may take one of the following values:
global, desk, page, screen, !desk, !page, or !screen. If global,
then all windows of the appropriate type (see the show and
dontshow options below) will be shown. If desk, then only those
windows on the current desk are shown. If page, then only those
windows on the current page are shown. If screen, then only those
windows on the current Xinerama screen are shown. !desk reverses
the sense of desk, displaying only those windows not on the
current desk. Likewise, !page shows only those windows not on the
current page and !screen shows only those windows not on the
current Xinerama screen. The default is page. If Xinerama is not
active or only a single screen is used, page and screen are
equivalent.

This configuration line is respected when MvwmIconMan is running
as well, the resolution is changed dynamically.

.IP "*MvwmIconMan: [id] Reverse \fIclass\fP"
Causes certain classes of buttons to have their relief lines reversed so that
up and down styles are reversed. This has no affect on flat buttons. The class
can be icon, normal or none. The default is none.

.IP "*MvwmIconMan: [id] SelectButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons when the
mouse is over them.

.IP "*MvwmIconMan: [id] SelectColorset \fIcolorset\fP"
Works like selectbutton but uses colorsets instead.  The style setting can
still only be applied with selectbutton.  See MvwmTheme.

.IP "*MvwmIconMan: [id] Shape \fIboolean\fP"
If \fITrue\fP, then use make the window shaped. Probably only useful if you
have multiple columns or rows. If MvwmIconMan wasn't compiled to support the
Shape extension, this generates an error message. When using shaped windows,
it's recommended that a mvwm style is made for MvwmIconMan that has no borders.
Otherwise, mvwm will get confused.

.IP "*MvwmIconMan: [id] Sort \fIvalue\fP"
If \fIname\fP, then the manager list is sorted by name. If \fInamewithcase\fP,
then it is sorted by name sensitive to case. If \fIid\fP, then
the manager list is sorted by the window id, which never changes after the
window is created. If \fIweighted\fP, then the manager list is sorted by
weight (see the description of \fIsortweight\fP below). Or it can be set to
\fInone\fP, which results in no sorting. Default is \fIname\fP.

.IP "*MvwmIconMan: [id] SortWeight \fIweight\fP \fIpattern-list\fP"
Assigns the specified \fIweight\fP to windows that match \fIpattern-list\fP.
The list is made up of patterns of the form \fItype=pattern\fP, where type
is one of \fIclass\fP, \fIresource\fP, \fItitle\fP, or \fIicon\fP, and pattern
is an expression of the same format used in the mvwm style command
(minimalistic shell pattern matching). Multiple sort weights can be given.
Each window is matched against the list of sort weights, in order, and is
given the weight from the first match. Lower-weighted windows are placed
first in the manager list. For example:
.EX
*MvwmIconMan: Sort       weighted
*MvwmIconMan: SortWeight 1 class=XTerm title=special*
*MvwmIconMan: SortWeight 10 class=XTerm
*MvwmIconMan: SortWeight 5
.EE
In this example, xterm windows whose titles start with "special" (weight 1)
are listed first, followed by everything but other xterms (weight 5), and the
other xterms (weight 10) are listed last. If no default weight (empty pattern
list) is given, the default weight is 0. Only relevant if the sort type is
set to \fIweighted\fP.

.IP "*MvwmIconMan: [id] Title \fItitle-string\fP"
Specifies the window title string for that manager window. \fITitlestring\fP
may either be a single word, or a string enclosed in quotes. The default is
"MvwmIconMan". This will be drawn in the title bar of the manager window, if
any, and in the title button, which is the button drawn when the manager is
empty.

.IP "*MvwmIconMan: [id] TitleButton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of the title button
(the button drawn when the manager is empty). The manager's title is drawn
in the title button.

.IP "*MvwmIconMan: [id] UseWinList \fIboolean\fP"
If \fItrue\fP, then honor the WinListSkip style flag. Otherwise, all windows
are subject to possible management according to the show and dontshow lists.

.PP
The two following options control which windows get handled by which
managers. A manager can get two lists, one of windows to show, and one of
windows to ignore. If only the \fIshow\fP list is given, then that manager
will show only the windows in the list. If only the \fIDontShow\fP list is
given, then the manager will show all windows except those in the list. If
both lists are given, then a window will be shown if it is not in the
\fIDontShow\fP list, and in the \fIShow\fP list. And finally, if neither list
is given, then the manager will handle all windows. Each list is made up of
patterns of the form \fItype=pattern\fP, where type is one of \fIclass\fP,
\fIresource\fP, \fItitle\fP, or \fIicon\fP, and pattern is an expression of
the same format used in the mvwm style command (minimalistic shell pattern
matching). Quotes around the pattern will be taken as part of the
expression. If a window could be handled by more than one manager, then the
manager with the lowest id gets it.

.IP "*MvwmIconMan: [id] Show \fIpattern list\fP"
If a window matches one of the patterns in the list, then it may be handled
by this manager.

.IP "*MvwmIconMan: [id] DontShow \fIpattern list\fP"
If a window matches one of the patterns in the list, then it may not be
handled by this manager.

.IP "*MvwmIconMan: [id] ShowTransient \fIboolean\fP"
Show transient windows in the list (default false).

.IP "*MvwmIconMan: [id] ShowOnlyIcons \fIboolean\fP"
Only iconified windows are shown if \fIboolean\fP is true.

.IP "*MvwmIconMan: [id] ShowNoIcons \fIboolean\fP"
Only windows that are not iconified are shown if \fIboolean\fP is true.

.IP "*MvwmIconMan: [id] ShowOnlyFocused \fIboolean\fP"
Only window with the focus is shown if \fIboolean\fP is true.


.PP
The following two options control tips.

.IP "*MvwmIconMan: [id] Tips \fIvalue\fP"
where \fIvalue\fP can be always, needed or false. Default is false,
no tips are displayed. With always, tips are enabled. With needed,
a tip is displayed only if either the button string is truncated or
the tip string is not equal to the button string.
This configuration line is respected when MvwmIconMan is running
as well.

.IP "*MvwmIconMan: [id] TipsDelays \fIdelay\fP [\fImappeddelay\fP]"
where \fIdelay\fP and \fImappeddelay\fP are time out values in milliseconds.
If no \fImappeddelay\fP is given \fIdelay\fP is assumed. Default is
1000 300. When the cursor is on a button, MvwmIconMan wait \fIdelay\fP
milliseconds before displaying the tip. In the case where a tip is
already mapped and the cursor goes to another button, MvwmIconMan
waits \fImappeddelay\fP milliseconds before displaying the new tip.

.IP "*MvwmIconMan: [id] TipsFont \fIfontname\fP"
Specifies the font to be used for tips. Default is the default mvwm
font.

.IP "*MvwmIconMan: [id] TipsColorset \fIcolorset\fP"
Specifies the colors for tips window. Default is colorset 0.
See  MvwmTheme.

.IP "*MvwmIconMan: [id] TipsFormat \fIformatstring\fP"
Similar to the Format option but for the tips window. The default is
the format string from the Format option.

.IP "*MvwmIconMan: [id] TipsBorderWidth \fIpixels\fP"
Specifies the border width (in pixels) of the tips window. Default is 1.


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.6 (not released yet)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-desktop-menu is re-written and provides better support of
    the XDG menu specification.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The optin "NoDeskLimitY" option of the GotoPage command did
    not work.

  - FvwmForm now honours a Xinerama screen specifier in its
    geometry.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Several possible format string attacks have been fixed.

-------------------------------------------------------------------
Changes in stable release 2.6.5 (20-Apr-2012)

* New features:

  - FvwmPager now wraps window names if SmallFont is set.

* Bug fixes:

  - Made signals registered with FVWM unblocking.
  - The "UnderMouse" option to PositionPlacement now honors the EWMH
    working area by default, if it's in use.
  - FvwmButtons handles the deleted window correctly for sub-windows.

-------------------------------------------------------------------
Changes in stable release 2.6.4 (01-Feb-2012)

* New features:

  - FvwmIconMan no longer allows for itself to be transient by using:

    Module FvwmIconMan Transient

    This is too ambiguous with valid module alias names, hence a
    transient FvwmIconMan must be created with "-Transient" as in:

    Module FvwmIconMan -Transient
  - New command InfoStore -- to store key/value pairs of information
    to relieve the burden of the SetEnv command.
  - Speed-up improvements for fvwm-menu-desktop when generating XDG
    menus by removing unnecessary stat(2) calls.
  - BusyCursor and CursorStyle have been set to make the cursor look
    like a dot during Read/PipeRead commands, by default.

* Bug fixes:

  - FvwmRearrange now understands the "ewmhiwa" option when placing
    windows.
  - Client gravity for subwindows is honoured for reparenting
    windows, such as with XEmbed, when changing the parent's
    geometry.
  - Conditional command processing of !Layer n, as in:

        Next (Xteddy, !Layer 4) Echo cuddles

    has now been fixed.
  - Fixed handling of swallowed windows for transient FvwmButtons'.
  - fvwm-menu-desktop now looks in /usr/share/applications for KDE
    legacy mode.
  - fvwm-config no longer accepts "--is-stable", "--is-released",
    "--is-final".

-------------------------------------------------------------------
Changes in stable release 2.6.3 (30-Sep-2011)

* New features:

  - New Style commands:

        TitleFormat
        IconTitleFormat

        Used to specify a window's visible name.  These new options
        deprecate the older styles:

        IndexedWindowName
        IndexedIconName
        ExactWindowName
        ExactIconName

  - New Resize argument to direction:  "automatic" which works out,
    automatically, which direction a window should be resized in
    based on which quadrant the pointer is in on the window -- this
    is best observed by the grid formed from "ResizeOutline", as the
    shape of this grid is used when calculating this.

    Furthermore, the existing option of "warptoborder" can be used in
    conjunction with "direction automatic" to warp the pointer to the
    appropriate window edge.

  - Couple of changes to the Move command:
        - The EWMH working area is now honoured by default.
        - To move a window ignoring the working area, the option
          "ewmhiwa" can be used, similar to how the Maximize
          command works.

  - New expansion placeholder $[pointer.screen] to return the screen number
    the pointer is on.

* Bug fixes:

    - A few minor ones; nothing user-visible per se.

-------------------------------------------------------------------

modules/MvwmIconMan/MvwmIconMan.c
/* -*-c-*- */
/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "config.h"

#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include "MvwmIconMan.h"
#include "readconfig.h"
#include "x.h"
#include "xmanager.h"

#include "libs/mvwmsignal.h"
#include "libs/Module.h"
#include "libs/FTips.h"
#include "libs/Parse.h"

char           *MyName;
FlocaleWinString *FwinString;
int             mods_unused = DEFAULT_MODS_UNUSED;

static RETSIGTYPE TerminateHandler(int);

char           *
copy_string(char **target, const char *src)
{
	int             len;
	if (src == NULL) {
		len = 0;
		if (*target) {
			Free(*target);
		}
		return *target = NULL;
	}

	len = strlen(src);
	ConsoleDebug(CORE, "copy_string: 1: 0x%lx\n",
	    (unsigned long) *target);

	if (*target) {
		Free(*target);
	}

	ConsoleDebug(CORE, "copy_string: 2\n");
	*target = mvwm_malloc((len + 1) * sizeof(char));
	strcpy(*target, src);
	ConsoleDebug(CORE, "copy_string: 3\n");
	return *target;
}

#ifdef TRACE_MEMUSE

long            MemUsed = 0;

void
Free(void *p)
{
	struct malloc_header *head = (struct malloc_header *) p;

	if (p != NULL) {
		head--;
		if (head->magic != MALLOC_MAGIC) {
			fprintf(stderr, "Corrupted memory found in Free\n");
			abort();
			return;
		}
		if (head->len > MemUsed) {
			fprintf(stderr, "Free block too big\n");
			return;
		}
		MemUsed -= head->len;
		Free(head);
	}
}

void
PrintMemuse(void)
{
	ConsoleDebug(CORE, "Memory used: %ld\n", MemUsed);
}

#else

void
Free(void *p)
{
	if (p != NULL) {
		free(p);
	}
}

void
PrintMemuse(void)
{
}

#endif

static          RETSIGTYPE
TerminateHandler(int sig)
{
	mvwmSetTerminate(sig);
	SIGNAL_RETURN;
}

void
ShutMeDown(int flag)
{
	ConsoleDebug(CORE, "Bye Bye\n");
	exit(flag);
}

void
DeadPipe(int nothing)
{
	(void) nothing;
	ShutMeDown(0);
	SIGNAL_RETURN;
}

static void
main_loop(void)
{
	fd_set          readset, saveset;
	fd_set_size_t   fd_width = x_fd;
	unsigned long   ms;
	struct timeval  tv;

	/*
	 * Calculate which is descriptor is numerically higher:
	 * this determines how wide the descriptor set needs to be ...
	 */
	if (fd_width < mvwm_fd[1]) {
		fd_width = mvwm_fd[1];
	}
	++fd_width;

	FD_ZERO(&saveset);
	FD_SET(mvwm_fd[1], &saveset);
	FD_SET(x_fd, &saveset);

	tv.tv_sec = 60;
	tv.tv_usec = 0;

	while (!isTerminated) {
		/*
		 * Check the pipes for anything to read, and block if
		 * * there is nothing there yet ...
		 */
		readset = saveset;

		if (mvwmSelect(fd_width, &readset, NULL, NULL, &tv) < 0) {
			ConsoleMessage
			    ("Internal error with select: errno=%s\n",
			    strerror(errno));
		} else {

			if (FD_ISSET(x_fd, &readset) || FPending(theDisplay)) {
				xevent_loop();
			}
			if (FD_ISSET(mvwm_fd[1], &readset)) {
				ReadMvwmPipe();
			}

		}
		if ((ms = FTipsCheck(theDisplay))) {
			tv.tv_sec = ms / 1000;
			tv.tv_usec = (ms % 1000) * 1000;
		} else {
			tv.tv_sec = 60;
			tv.tv_usec = 0;
		}
	}	/* while */
}

int
main(int argc, char **argv)
{
	int             i;

#ifdef HAVE_LIBEFENCE
	extern int      EF_PROTECT_BELOW, EF_PROTECT_FREE;

	EF_PROTECT_BELOW = 1;
	EF_PROTECT_FREE = 1;
#endif

	FlocaleInit(LC_CTYPE, "", "", "MvwmIconMan");

	OpenConsole(OUTPUT_FILE);

	init_globals();
	init_winlists();

	MyName = GetFileNameFromPath(argv[0]);

	if (argc >= 7) {
		if (strcasecmp(argv[6], "-Transient") == 0) {
			globals.transient = 1;

			/*
			 * Optionally accept an alias to use as the transient
			 * * MvwmIconMan instance.
			 */
			if (argv[7])
				MyName = argv[7];
		} else {
			MyName = argv[6];
		}
	}
	ModuleLen = strlen(MyName) + 1;
	Module = mvwm_malloc(ModuleLen + 1);
	*Module = '*';
	strcpy(Module + 1, MyName);

	if (argc < 6) {
		fprintf(stderr,
		    "%s version %s should only be executed by mvwm!\n",
		    MyName, VERSION);
		ShutMeDown(1);
	}
	mvwm_fd[0] = atoi(argv[1]);
	mvwm_fd[1] = atoi(argv[2]);
	init_display();

#ifdef HAVE_SIGACTION
	{
		struct sigaction sigact;

		sigemptyset(&sigact.sa_mask);
		sigaddset(&sigact.sa_mask, SIGPIPE);
		sigaddset(&sigact.sa_mask, SIGINT);
		sigaddset(&sigact.sa_mask, SIGHUP);
		sigaddset(&sigact.sa_mask, SIGTERM);
		sigaddset(&sigact.sa_mask, SIGQUIT);
#ifdef SA_RESTART
		sigact.sa_flags = SA_RESTART;
#else
		sigact.sa_flags = 0;
#endif
		sigact.sa_handler = TerminateHandler;

		sigaction(SIGPIPE, &sigact, NULL);
		sigaction(SIGINT, &sigact, NULL);
		sigaction(SIGHUP, &sigact, NULL);
		sigaction(SIGTERM, &sigact, NULL);
		sigaction(SIGQUIT, &sigact, NULL);
	}
#else
	/*
	 * We don't have sigaction(), so fall back to less robust methods.
	 */
#ifdef USE_BSD_SIGNALS
	mvwmSetSignalMask(sigmask(SIGPIPE) |
	    sigmask(SIGINT) |
	    sigmask(SIGHUP) | sigmask(SIGTERM) | sigmask(SIGQUIT));
#endif
	signal(SIGPIPE, TerminateHandler);
	signal(SIGINT, TerminateHandler);
	signal(SIGHUP, TerminateHandler);
	signal(SIGTERM, TerminateHandler);
	signal(SIGQUIT, TerminateHandler);
#ifdef HAVE_SIGINTERRUPT
	siginterrupt(SIGPIPE, 0);
	siginterrupt(SIGINT, 0);
	siginterrupt(SIGHUP, 0);
	siginterrupt(SIGTERM, 0);
	siginterrupt(SIGQUIT, 0);
#endif
#endif

	read_in_resources();
	FlocaleAllocateWinString(&FwinString);

	for (i = 0; i < globals.num_managers; i++) {
		X_init_manager(i);
	}

	assert(globals.managers);

	SetMessageMask(mvwm_fd,
	    M_CONFIGURE_WINDOW | M_RES_CLASS | M_RES_NAME | M_ADD_WINDOW |
	    M_DESTROY_WINDOW | M_ICON_NAME | M_DEICONIFY | M_ICONIFY |
	    M_ICON_LOCATION | M_END_WINDOWLIST | M_NEW_DESK | M_NEW_PAGE |
	    M_FOCUS_CHANGE | M_WINDOW_NAME | M_CONFIG_INFO | M_SENDCONFIG |
	    M_VISIBLE_NAME | M_MINI_ICON | M_STRING | M_WINDOWSHADE |
	    M_DEWINDOWSHADE);
	/*
	 * extended messages
	 */
	SetMessageMask(mvwm_fd, MX_VISIBLE_ICON_NAME | MX_PROPERTY_CHANGE);

	SendText(mvwm_fd, "Send_WindowList", 0);

	/*
	 * tell mvwm we're running
	 */
	SendFinishedStartupNotification(mvwm_fd);

	/*
	 * Lock on send only for iconify and deiconify (for NoIconAction)
	 */
	SetSyncMask(mvwm_fd, M_DEICONIFY | M_ICONIFY);
	SetNoGrabMask(mvwm_fd, M_DEICONIFY | M_ICONIFY);

	main_loop();

	return 0;
}
